### PR TITLE
[Storage][Sharding] Cache crypto hash in StateKey.

### PR DESCRIPTION
**Pull request by grao1991**
### Description

1. Avoid recalculation.
2. Will be used as sharding key.

### CODE REVIEW

**Review by msmouse — APPROVED** [review]

File: types/src/state_store/state_key.rs [hunk 1]
```diff
@@ -2,22 +2,31 @@
...
 use aptos_crypto_derive::CryptoHasher;
 use num_derive::{FromPrimitive, ToPrimitive};
 use num_traits::FromPrimitive;
-use serde::{Deserialize, Serialize};
-use std::convert::TryInto;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{convert::TryInto, ops::Deref};
 use thiserror::Error;
 
+#[derive(Clone, Debug, CryptoHasher, Eq, PartialEq, Ord, PartialOrd, Hash)]
```
Comment (gelash) [inline types/src/state_store/state_key.rs:17]: Should it derive it if it implements CryptoHasher explicitly (as it should)?

Also, I was always confused of the fact that it also derives Hash, and the data-structures can be using that (e.g. the non-cryptographic hash), which could also bypass the current optimization if this isn't mistaken?

Comment (grao1991) [reply types/src/state_store/state_key.rs:17]: The line 139 requires this derive (although I don't think it is used).

This PR is not trying to solve the "Dashmap compute (non-crypto) hash twice" issue. I need to double check whether caching it is a good idea or we should go with a different approach. (e.g. use the first byte of the crypto hash as the shard num, etc.)

For the reference, based on my benchmark, for StateKey, in average, CryptoHash takes ~1ms, Hash takes ~90ns.

Comment (gelash) [reply types/src/state_store/state_key.rs:17]: I see re:Hash, and indeed makes sense that data-structures would not use cryptographic hash, but we could use a similar approach (and should be significantly simpler to add) later to optimize the normal hash as well, right? Just checking my understanding of what's going on.

re:derive, I think it's confusing to have it there (as a reader like myself would falsely assume it wouldn't also be explicitly implemented), but I see line:139 and what you mean. I think you might be able to invoke the 'define_hasher' macro (but it's not exported, so maybe inside hash.rs) and create a fake StateKeyHasher without deriving. There is some context in the aptos_crypto hash.rs documentation.

Comment (grao1991) [reply types/src/state_store/state_key.rs]: re hash: right

re derive: done, removed the derive, and added a dummy hasher in hash.rs.

Comment (gelash) [reply types/src/state_store/state_key.rs]: Thanks!

**Review by msmouse — CHANGES_REQUESTED** [review]
per offline discussion, let's make it lazy?

**Comment by grao1991** [comment]
> per offline discussion, let's make it lazy?

Done

File: types/src/state_store/state_key.rs [hunk 2]
```diff
@@ -2,22 +2,41 @@
...
+#[cfg_attr(any(test, feature = "fuzzing"), derive(proptest_derive::Arbitrary))]
+pub struct StateKey {
+    inner: StateKeyInner,
+    #[derivative(
+        Hash = "ignore",
+        Ord = "ignore",
+        PartialEq = "ignore",
+        PartialOrd = "ignore"
+    )]
+    #[cfg_attr(any(test, feature = "fuzzing"), proptest(value = "OnceCell::new()"))]
```
Comment (msmouse) [inline types/src/state_store/state_key.rs:30]: 🎉

**Review by msmouse — APPROVED** [review]

**Review by areshand — APPROVED** [review]

**Pull request merged**